libs/dexie-react-hooks/src/useSuspendingObservable.ts

```diff
@@ -0,0 +1,149 @@
+import * as React from 'react';
+import { AnySubscription, InteropableObservable } from './useObservable';
+import { usePromise } from './usePromise';
+
+/**
+ * Subscribes to an observable and returns the latest value.
+ * Suspends until the first value is received.
+ *
+ * Calls with the same cacheKey will use the same observable.
+ * cacheKey must be globally unique.
+ */
+export function useSuspendingObservable<T>(
+  getObservable: (() => InteropableObservable<T>) | InteropableObservable<T>,
+  cacheKey: React.DependencyList
+): T {
+  let observable: InteropableObservable<T>;
+  for (const [key, val] of OBSERVABLES.entries()) {
+    if (
+      key.length === cacheKey.length &&
+      key.every((k, i) => Object.is(k, cacheKey[i]))
+    ) {
+      observable = val;
+      break;
+    }
+  }
+  //@ts-ignore (because observable might be undefined here)
+  if (!observable) {
+    observable = typeof getObservable === 'function'
+      ? getObservable()
+      : getObservable;
+    OBSERVABLES.set(cacheKey, observable);
+  }
+  // At this point, observable is always set.
+
+  const incrementRef = () => {
+    const timeout = TIMEOUTS.get(observable);
+    if (timeout != null) clearTimeout(timeout);
+
+    let refCount = REF_COUNTS.get(observable) ?? 0;
+    refCount += 1;
+    REF_COUNTS.set(observable, refCount);
+  };
+
+  const decrementRef = () => {
+    let refCount = REF_COUNTS.get(observable)!;
+    refCount -= 1;
+    REF_COUNTS.set(observable, refCount);
+
+    if (refCount > 0) return;
+
+    const timeout = setTimeout(() => {
+      for (const [key, val] of OBSERVABLES.entries()) {
+        if (val === observable) {
+          OBSERVABLES.delete(key);
+          break;
+        }
+      }
+      PROMISES.delete(observable);
+      VALUES.delete(observable);
+      TIMEOUTS.delete(observable);
+      REF_COUNTS.delete(observable);
+    }, 1000);
+    TIMEOUTS.set(observable, timeout);
+  };
+
+  let promise: Promise<T> | undefined = PROMISES.get(observable);
+  if (!promise) {
+    promise = new Promise<T>((resolve, reject) => {
+      if (VALUES.has(observable)) {
+        resolve(VALUES.get(observable)!);
+        return;
+      }
+
+      incrementRef();
+
+      let calledSynchronously = false;
+      let sub: AnySubscription;
+      sub = observable.subscribe(
+        (val) => {
+          resolve(val);
+          VALUES.set(observable, val);
+          if (!sub) calledSynchronously = true;
+          else unsub(sub);
```

Comment:
@phaux @dfahlander I've noticed an issue with `useSuspendingLiveQuery` where the `querier` function is executed twice even though the dependencies and underlying data doesn't change.

Upon further investigation, calling `subscribe` on an observable appears to re-executes the query. Is this correct? I'm surprised the observable doesn't return the existing value when re-subscribing.

Since we subscribe both here and in the `useEffect` block below this is causing the `querier` to execute twice, once for each `subscribe` call.

I want to verify my assumptions here before creating an issue for this.

Reply:
That seems correct. it's something that could be improved.